facebook/docusaurus
# feat(content-docs): draft docs excluded from build & sidebars

## Motivation

The `plugin-content-docs` plugin does not currently provide a way for documents to be worked on while remaining hidden to visitors of the site. Its sibling plugin, `plugin-content-blog` provides this via the [`draft`](https://docusaurus.io/docs/api/plugins/@docusaurus/plugin-content-blog#draft) frontmatter property, which excludes the blog post from builds. This PR aims to reach parity with the blog plugin and introduce this functionality as well.

### `draft`

Use case:
  - Document is actively being worked on, but is not ready for public consumption, and can be merged into the main/deployed Git branch without risk of leaking in production builds
  
Marking a document with `draft: true`...

- Excludes it entirely from production builds
  - Remains visible in development
  - When built, accessing the URL yields a 404. There was [discussion](https://github.com/facebook/docusaurus/issues/3417#issuecomment-689139080) around allowing this to be customized such that instead of displaying a 404 you could display a "work in progress" message, however that behavior is inconsistent with the blog equivalent so I did not go that route. We could consider something like this in the future with an "unlisted" type draft status.
- Excludes it from sidebars in production builds
  - This includes sidebars both manually and automatically generated
- Excludes it from automatically generated index pages in production builds

### Have you read the [Contributing Guidelines on pull requests](https://github.com/facebook/docusaurus/blob/main/CONTRIBUTING.md#pull-requests)?
Yes

## Test Plan

First, create a new document with any content you wish. Add this document to a sidebar manually (or set up a sidebar to auto generate).  This document should be visible in development and production builds.

To test `draft`...

1. Add `draft: true` to the document's frontmatter, and save it.
2. Start the development server. The expected result is this document remains accessible both directly and via its place in a sidebar (both manual and generated).
3. Generate a production build and serve the static site. The expected result is this document is **not** accessible directly and is not visible in any sidebar (both manual and generated). Additionally, it should not show up in auto-generated index pages.

## Related PRs

Closes #3417

---

🗺 I looked into how this impacts the sitemaps plugin, and for `draft` no extra work is necessary as these documents are already filtered out in `routes.ts`, preventing their page from being created and passing the route down to the plugin.
